dipy/workflows/multi_io.py

```diff
@@ -253,3 +253,10 @@ def __iter__(self):
         IO = np.concatenate([I, O], axis=1)
         for i_o in IO:
             yield i_o
+
+    @staticmethod
+    def file_existence_check(args):
+        input_args = list(args)
+        for input_path in sorted(input_args):
+            if len(glob(input_path)) == 0:
```

Comment:
This condition works but it is not clear. Can you change it and use something more explicit like:
`if not os.path.isfile(input_path) or not os.path.isdir(input_path)`

Reply:
@serge thank you for checking out the code. So you are right the way I am doing it is not intuitive but below are  the reasons why I am doing it this way, So I am writing the code to do what you suggested,

        for input_path in sorted(input_args):
            if os.path.isdir(input_args):
                if not os.path.isdir(input_path):
                    err = True
            elif os.path.isfile(input_path):
                if not os.path.isfile(input_path):
                    err = True
            if err:
                raise IOError('File not found: ' + input_path)

You can see that O have to separately check for the file and a directory and accordingly decide what is missing, this increases the number of IF-ELSE conditions for this small check and since this check is performed for all the workflows so this will cause the slowdown when the workflows are executed multiple times.

So, what i did is to keep the default glob object and check if it's length is 0 so that there is just one IF condition to check.

Please tell if there is a better way to implement this check that I can use? otherwise, I will implement the above solution which is more intuitive but a bit slower.

Thanks.

Reply:
I will wait for your suggestion and only commit the changes after that.

Reply:
Hey serge, thanks for your suggestion, here is where I think this check can fail,

We know that currently, the workflows work for both absolute file/directory paths and also for wildcards, in case the user gives absolute file or directory names then this is fine but when the user supplies a wildcard for example, in the apply_affine workflow to provide all the input files at once,
For example:
dipy_apply_affine *input --affine=affine.txt

In this case, you can see the above check will always fail since this check will simply look for file *input but actually, it can be a file called input1.nii.gz or input2.nii.gz etc. and so it will always fail with the file not found.

I tested this behavior when I wrote this check and realized that a simple condition won't be enough. So, to keep the code to the minimum, I used the len(glob(input_path)) so that for both absolute file names and wildcard based file names the check can work because the glob will automatically expand the wildcard to the actual file names.

I realized this when I was starting to make the changes again based on your suggestion.

Let me know what you think about it?

Reply:
Ok, I see, I totally forgot the wildcards. Thank you for all this explanation. I understand why you use `glob`.

No need to change, You can let it as it is.